graphql-kotlin-federation/src/main/kotlin/com/expediagroup/graphql/federation/tracing/FederatedTracingInstrumentation.kt

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright 2019 Expedia, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.expediagroup.graphql.federation.tracing
+
+import com.expediagroup.graphql.exceptions.GraphQLKotlinException
+import com.google.protobuf.Timestamp
+import graphql.ExecutionResult
+import graphql.ExecutionResultImpl
+import graphql.GraphQLError
+import graphql.GraphqlErrorBuilder
+import graphql.execution.DataFetcherResult
+import graphql.execution.ExecutionPath
+import graphql.execution.ExecutionStepInfo
+import graphql.execution.instrumentation.InstrumentationContext
+import graphql.execution.instrumentation.InstrumentationState
+import graphql.execution.instrumentation.SimpleInstrumentation
+import graphql.execution.instrumentation.SimpleInstrumentationContext.whenCompleted
+import graphql.execution.instrumentation.parameters.InstrumentationCreateStateParameters
+import graphql.execution.instrumentation.parameters.InstrumentationExecutionParameters
+import graphql.execution.instrumentation.parameters.InstrumentationFieldFetchParameters
+import graphql.language.Document
+import graphql.language.SourceLocation
+import graphql.parser.InvalidSyntaxException
+import graphql.schema.GraphQLTypeUtil.simplePrint
+
+import mdg.engine.proto.Reports
+import java.time.Instant
+import java.util.ArrayList
+import java.util.Base64
+import java.util.concurrent.CompletableFuture
+import kotlin.collections.LinkedHashMap
+
+class FederatedTracingInstrumentation : SimpleInstrumentation() {
+    companion object {
+        const val HEADER_NAME: String = "apollo-federation-include-trace"
+        const val HEADER_VALUE: String = "ftv1"
+        const val EXTENSION_KEY: String = "ftv1"
+    }
+
+    override fun createState(parameters: InstrumentationCreateStateParameters): InstrumentationState? {
+        val context = parameters.executionInput.context
+        if (context is HttpRequestHeaders) {
```

Comment:
This puts a requirement on the context object types. If we are going to go with this route we should create something very generic that has access to request and response objects instead.